include/oneapi/dpl/pstl/utils_ranges.h

```diff
@@ -478,7 +478,7 @@ struct permutation_view_simple<_Source, _M, ::std::enable_if_t<oneapi::dpl::__in
     //TODO: to be consistent with C++ standard, this Idx should be changed to diff_type of underlying range
     template <typename Idx>
     auto
-    operator[](Idx __i) const
+    operator[](Idx __i) const -> decltype(__src[__map_fn(__i)])
```

Comment:
Why is not the compiler able to deduce the return value type from the return statement?

Reply:
I believe that the compiler is deducing and returning the `value_type` of the base iterator without this guide, and we need the `operator[]` to return the `reference_type`, the same as the return of the `operator[]` for the base iterator.

I'm not sure how the compiler makes this decision, but this guide makes it explicit and resolves the error.  I can look up the rules for auto return in this scenario to try to bring more clarity here.

Reply:
Following up, I believe that without any trailing `-> ...`, auto uses the normal template argument deduction, which will result in the **referenced** type being selected, not the reference type. https://en.cppreference.com/w/cpp/language/template_argument_deduction#Deduction_from_a_function_call (3)

I've adapted the example from [here](https://dice-research.org/news/2022-04-08_cpp_return_type_auto/) to explore this topic a bit below:
https://godbolt.org/z/f8dr5eEea


It seems a more concise way of accomplishing this would be to just use `decltype(auto)` as the return with no trailing definition of the type.  This skips the template argument deduction entirely and is a placeholder which uses the decltype of the returned expression.  I don't have a preference between the two options.

Reply:
May be better to use `-> decltype(__src[__map_fn(__i)])` as you did because this approach already used in our code and it has been used here before https://github.com/oneapi-src/oneDPL/pull/1254

Reply:
> this would be to just use `decltype(auto)`

I remember that Ruslan recommended to use this way.

Reply:
Considering it further, I like `decltype(auto)` better as it removes a redundant bit of code and is more maintainable.  I've made the change.